### PR TITLE
Work Packets

### PR DESCRIPTION
**Pull request by wenyuzhao**
This PR refactors the mmtk-core to use work packets for GC. Related to Issue #48.

**_~~This is a draft PR with a prototype implementation.~~ Discussions are welcome._**

* **JikesRVM Binding PR:** https://github.com/mmtk/mmtk-jikesrvm/pull/30
* **JikesRVM PR:** https://github.com/mmtk/jikesrvm/pull/4
* **OpenJDK Binding PR:** https://github.com/mmtk/mmtk-openjdk/pull/25
* **OpenJDK PR:** https://github.com/mmtk/openjdk/pull/4

This PR also fixes Issue #155.

# Work-Packet System Design

The work packet system consists of a scheduler, a set of worker threads, and a set of work packets.

### Scheduler

The scheduler contains a list of work buckets, each has its own open condition. The `unconstrained` bucket is always open and works in this bucket are schedulable at any time (even within mutator time). The other buckets are closed by default and will be open if certain open conditions are met.

Part of the scheduler code is run in a coordinator thread (see [coordinator-thread](#coordinator-thread) section).

The `Scheduler::poll` method is used by worker threads to pop a schedulable work packet from one of the (open) work buckets. 

##### Work bucket

Each work bucket is a priority queue wrapped with a spin-lock. After a bucket is drained, a message is sent to the coordinator thread through a _channel_ to notify the coordinator thread to update buckets' status.

If a new work packet is added to a bucket, _one worker thread_ will wake up (through a condition variable).

##### Coordinator thread

The coordinator thread is responsible for:
1. Wait for GC request, and schedule GCs if required.
2. Execute works that have to be done in the coordinator thread. e.g. stop/resume mutators -- because OpenJDK requires the thread that resumes mutators to be the same one for stop mutators.
3. Check buckets' open conditions and open buckets if possible. Also, waits up _all the workers_ if a new bucket is open.
4. Check if the GC is completed, if so, return to _Step#1_ and wait for future GC requests.

### Work packets

Each work packet implements the `Work<C: Context>` trait. Work packets will then be sent across worker threads and the global scheduler as a trait object. The use of trait objects here meets the requirement of "_self describing_" (data + function pointer -- see Issue #48).

```rust
trait Work<C: Context> {
    fn do_work(&mut self, worker: &'static mut Worker<C>, context: &'static C);
}
```

The `context` parameter is a global variable that contains the data required for the execution of the whole system. For a work-packet system that does GC (e.g. mmtk-core), the context is `&'static MMTK<VM>`.

Under the context of GC, all GC-related work packets implement the `trait GCWork<VM: VMBinding>: Work<MMTK<VM>>` trait. Common GC works are defined in [gc_works.rs](https://github.com/mmtk/mmtk-core/blob/work-packets/src/scheduler/gc_works.rs). Each GC plan and VM binding also define their own GC works.

### Worker thread

A worker thread runs in an infinite loop and keeps polling new schedulable works from the global scheduler. Under most cases, the worker will busy waiting for the spin-locks of the buckets (see [work-buckets](#work-buckets) section) to get schedulable works. If there are no more schedulable works, the worker thread will then sleep by using a mutex and a condition variable. A slept worker will wake up by the coordinator if a new bucket is open (see [coordinator-thread](#coordinator-thread) section) or by a work bucket if a new work is added to the bucket (see [work-bucket](#work-bucket) section).

### Generality

the work-packet system is designed with generality in mind. For an example of using the system for other purposes, see [this test](https://github.com/mmtk/mmtk-core/blob/work-packets/tests/scheduler.rs).

# GC Work Packets

Two collectors are implemented using this work packets system:
* SemiSpace
* GenCopy (generation semispace without barriers)
  * -- This GC is just an example of how to implement a generational GC with code reuse under the work packets system. Will be removed before the merge of this PR.

Here list all the GC work packets used by the above two GCs

### Common GC works

* **`trait ProcessEdgesWork: GCWork<VM>`** Each GC defines a new edge-processing work that implements this trait for different behaviours of objects tracing.

##### Schedulable at any time (no constraints)
* **`StopMutators<ScanEdges: ProcessEdgesWork>`**  Stop all mutators and schedules `ScanStackRoot` for each mutator immediately after it is yielded. 
* **`ScanStackRoot<Edges: ProcessEdgesWork>`** Scan the stack of one mutator, and schedules instances of `ProcessEdgesWork` for tracing the roots.

##### [Prepare-Bucket] Open after all mutators are paused
* **`Initiate<P: Plan>`**  Set the collection kind and set the GC status to `GcStatus::GcPrepare`
* **`Prepare<P: Plan>`**  Global preparation
* **`PrepareMutator<P: Plan>`**  Prepare one mutator
* **`ScanVMSpecificRoots<Edges: ProcessEdgesWork>`** Notify each vm binding to create their non-stack roots scanning works. See [openjdk-specific-works](#openjdk-specific-works) section.
  * **`ScanStackRoots<Edges: ProcessEdgesWork>`** In case some VMs does not support scan stack roots mutator-by-mutator (or hard to implement), this work is used to scan all the stack roots as a whole.

##### [Closure-Bucket] Open after _[Prepare-Bucket]_ is drained
* **`trait ProcessEdgesWork: GCWork<VM>`** Process a list of root or non-root edges. The behaviour of edge processing (could be marking or evacuating and then updating slots), is GC-specific. Produces several `ScanObjects` works for recursive edge scanning.
* **`ScanObjects<Edges: ProcessEdgesWork>`** Scan the fields of a list of objects. Produce instances of `ProcessEdgesWork` that contains the discovered edges.

##### [Release-Bucket] Open after _[Closure-Bucket]_ is drained
* **`Release<P: Plan>`**  Global release works
* **`ReleaseMutator <P: Plan>`**  Release a mutator

##### [Final-Bucket] Open after _[Release-Bucket]_ is drained
* **`ResumeMutators`**  Send a message to the coordinator thread to resume all the mutators.

##### Works directly sent to each worker-local work queue instead of the global buckets.
* **`PrepareCollector`**  Prepare a worker thread
* **`ReleaseCollector`**  Release a worker thread

### SemiSpace works
* **`SSProcessEdges<VM: VMBinding>`** implements `ProcessEdgesWork`. For each edge, calls `tospace.trace_object` or `fromspace.trace_object`, depending which space the edge points to. The `**space.trace_object` methods race to copy the object, and push the forwarded object to a buffer for recursive edge scanning. A `ScanObjects` work will be created if the buffer is filled.

### GenCopy works
* **`GenCopyMatureProcessEdges<VM: VMBinding>`** same as `SSProcessEdges`, but also evacuates live nursery objects to `tospace`.
* **`GenCopyNurseryProcessEdges<VM: VMBinding>`** same as `GenCopyMatureProcessEdges`, but evacuate nursery objects only. Because this implementation of GenCopy does not use barriers, all mature objects are visited and marked using a side mark table.

### OpenJDK specific works

There are 12 different types of vm-specific root scanning works for OpenJDK. Each of them will produce instances of `ProcessEdgesWork`.

* **`ScanUniverseRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanJNIHandlesRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanObjectSynchronizerRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanManagementRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanJvmtiExportRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanAOTLoaderRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanSystemDictionaryRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanCodeCacheRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanStringTableRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanClassLoaderDataGraphRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanWeakProcessorRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**
* **`ScanVMThreadRoots<E: ProcessEdgesWork<VM=OpenJDK>>`**

# Statistics

Within the call of `harness_end`, the scheduler reports:
* Count of each type of work packet
* Geomean execution time of each type of work packet
* Total count of all work packets
* Geomean/min/max execution time of all work packets

The statistics are collected by each worker thread and saved in worker-local buffers. When reporting results, the scheduler will get data from each worker thread and summarize them all.

# TODO

* [x] Fix nogc build error
* [x] GenCopy using barriers (concurrent work packet)
* [x] Fix the buggy side mark-table for GenCopy
* [x] Sanity GC support
* [x] JikesRVM / V8 support
* [x] Performance comparison

# Future work

* [ ] Statistics of worker idle / suspend time
* [ ] Further optimizations (reduce the use of heavy-locks / channels?)

### CODE REVIEW

File: src/plan/barriers.rs
Comment (wenyuzhao) [reply src/plan/barriers.rs]: Right, currently the assumption of nursery Copyspace is unnecessary. I've added a generic argument `S: Space<VM>` to `FieldRememberingBarrier` and removed the `cfg` guards.

File: src/plan/barriers.rs
Comment (wenyuzhao) [reply src/plan/barriers.rs]: Done. It's now renamed to `modified_nodes` / `modified_edges` and live in a separate struct `ModBuffer`.

File: src/plan/gencopy/gc_works.rs
Comment (wenyuzhao) [reply src/plan/gencopy/gc_works.rs]: The assumption is `OVERWRITE_REFERENCE == true` by default. I've removed this if-statement since we guarantee we need reference overwriting here.

File: src/plan/barriers.rs
Comment (wenyuzhao) [reply src/plan/barriers.rs]: Done.

File: src/plan/gencopy/global.rs
Comment (wenyuzhao) [reply src/plan/gencopy/global.rs]: Done.

File: src/plan/gencopy/mutator.rs
Comment (wenyuzhao) [reply src/plan/gencopy/mutator.rs:41]: `gencopy` is only for openjdk and cannot run on jikesrvm (no barriers support). It is correct for openjdk since openjdk has no immortal/code spaces. I'll leave a TODO comment here about this.

File: src/plan/global.rs
Comment (wenyuzhao) [reply src/plan/global.rs:98]: This is for simplifying the code for defining structs with generic arguments (especially work-packets). For example, work packet `Prepare<VM: VMBinding, P: Plan<VM>>` can be simplified to `Prepare<P: Plan>`.

File: src/plan/transitive_closure.rs
Comment (wenyuzhao) [reply src/plan/transitive_closure.rs:12]: I think we can remove `TransitiveClosure`, and `TraceLocal` as well. Probably in another PR. There are already too many changes in this PR.

File: src/policy/space.rs
Comment (wenyuzhao) [reply src/policy/space.rs]: Done.

File: src/scheduler/context.rs
Comment (wenyuzhao) [reply src/scheduler/context.rs:8]: Done. MMTk-related implementation and type definitions are moved to `mmtk_context.rs`, with comments to explain the design,

File: src/scheduler/gc_works.rs
Comment (wenyuzhao) [reply src/scheduler/gc_works.rs:40]: Done.

File: src/scheduler/gc_works.rs
Comment (wenyuzhao) [reply src/scheduler/gc_works.rs]: Yes, it's correct because this work-packet does not outlive a GC. I've added a comment here to explain this.

Probably we need a `<'gc>` lifetime parameter in the future?

File: src/scheduler/gc_works.rs
Comment (wenyuzhao) [reply src/scheduler/gc_works.rs]: Done. It's in BasePlan now.

File: src/scheduler/gc_works.rs
Comment (wenyuzhao) [reply src/scheduler/gc_works.rs:276]: This is for fast pointer dereferencing. We need to dereference this pointer once for each object copy (to get work-local allocators), which is highly frequent for semispace.

We can't use `&'static mut GCWorker` because we don't know which one is the execution worker when this work-packet is created. But using `Option<&'static mut GCWorker>` will introduce one branch, which may down the performance if we have frequent pointer dereferencing (or probably not too much effect?).

Probably I can use `MaybeUninit` or `UnsafeCell` instead of using raw pointers. Or just use `Option<&'static mut GCWorker>`.

File: src/scheduler/scheduler.rs
Comment (wenyuzhao) [reply src/scheduler/scheduler.rs:23]: For _pmd_ benchmark in _dacapo-9.12_, there will be over 50 thousand work-packets being created for each iteration. Doing 50K times of WorkBucket lookups may be a bad idea, even when using `EnumMap`s. But the synchronization cost for adding a work-packet is likely to be higher than doing a bucket lookup. If this is the case, we can then switch to `Map<Stage, WorkBucket>`.

File: src/scheduler/stat.rs
Comment (wenyuzhao) [reply src/scheduler/stat.rs:17]: Done.

File: src/scheduler/work.rs
Comment (wenyuzhao) [reply src/scheduler/work.rs:23]: Thanks for pointing this out. I did some tests of address comparison of boxed structs / trait-objects. It turns out that my assumption here is incorrect. If a struct has no fields, then there will be no heap allocation. The boxed value will hold a fake pointer (e.g. 0x01) instead of a pointer into the heap.

The purpose of this address comparison is for fast work-packet comparison. Otherwise comparing (or calculating hash value for) work-packets with large data buffers will be extremely slow. I'll add a unique work-id for each instance of work-packet and use that for fast comparison.

File: src/scheduler/work_bucket.rs
Comment (wenyuzhao) [reply src/scheduler/work_bucket.rs]: Ah, this code is incorrect. ` &self.work.as_ref() as *const _` has type `*const *const dyn T` which is likely to be a stack pointer... Thanks for pointing this out.

File: src/scheduler/work_bucket.rs
Comment (wenyuzhao) [reply src/scheduler/work_bucket.rs]: Removed

File: src/scheduler/gc_works.rs
Comment (qinsoon) [reply src/scheduler/gc_works.rs:276]: Probably just add the above as comments so we know the reason behind this pointer use.

File: src/scheduler/worker.rs
Comment (wenyuzhao) [reply src/scheduler/worker.rs:16]: WorkerLocal holds the GC-specific thread-local data. For MMTk, it is `Plan::CopyContext`, and holds the local copy allocators.

File: src/util/heap/monotonepageresource.rs
Comment (wenyuzhao) [reply src/util/heap/monotonepageresource.rs:58]: This overrides `CommonPageResource::reserve_pages` to make MonotonePageResource (with contiguous memory) to be able to have non-zero metadata pages.

File: src/util/heap/pageresource.rs
Comment (wenyuzhao) [reply src/util/heap/pageresource.rs]: Done.

File: src/util/opaque_pointer.rs
Comment (wenyuzhao) [reply src/util/opaque_pointer.rs]: It is used for jikesrvm binding to get the internal data from a tls pointer. I removed this method so that mmtk-core will not be read/write to `OpaquePointer`. In jikesrvm binding, it transmutes the `OpaquePointer` to `Address` to read tls data.

File: src/util/heap/monotonepageresource.rs
Comment (qinsoon) [reply src/util/heap/monotonepageresource.rs:58]: I dont get this. I meant this function looks the same as https://github.com/mmtk/mmtk-core/blob/1c806b4b6062ddce8b1036776f2628eb811decd7/src/util/heap/pageresource.rs#L131. If we use that one, we do not need to make `reserved` as `pub` and can remove this function here.

File: src/scheduler/gc_works.rs
Comment (wenyuzhao) [reply src/scheduler/gc_works.rs:380]: The process-edges work is defined by each gc. For semispace it defines `SSProcessEdges` and implements `ProcessEdgesWork`.

The code here uses the `#![feature(specialization)]` feature to provide a default `GCWork` implementation for all process-edges works. This reduces code duplication. All `ProcessEdgesWork`s automatically implements `GCWork`.

But a gc algorithm can still manually implement `GCWork` for its process-edges work again if the algorithm wants to do some extra things in the `GCWork::do_work` method.

File: src/scheduler/gc_works.rs
Comment (wenyuzhao) [reply src/scheduler/gc_works.rs:319]: Not sure if this is an established pattern (or even a good pattern). But this can reduces the frequent use of `self.base().xxx` and can simplifies the code a lot. Probably I'll change to `self.base()` to keep the style aligned with the `Plan`/`BasePlan` code.

File: src/scheduler/gc_works.rs
Comment (qinsoon) [reply src/scheduler/gc_works.rs:380]: I see. I didn't know this feature, and couldn't find anything useful by just searching 'rust default fn'. The feature is described here: https://github.com/rust-lang/rfcs/blob/master/text/1210-impl-specialization.md

File: src/scheduler/gc_works.rs
Comment (qinsoon) [reply src/scheduler/gc_works.rs:319]: I am not feeling negative about it. It just made me wonder why it derefs to a base type. Once I figured out it was used to access fields from the base type, it makes sense.

File: src/util/heap/monotonepageresource.rs
Comment (wenyuzhao) [reply src/util/heap/monotonepageresource.rs:58]: Right. I've fixed that. It now calls `self.reserve(...)` instead of accessing the field.

Comment (wenyuzhao) [reply src/util/heap/monotonepageresource.rs:58]: But the function should not be moved to CommonPageResource. It is a method from `trait PageResource` defined at https://github.com/mmtk/mmtk-core/blob/master/src/util/heap/pageresource.rs#L28

Comment (qinsoon) [reply src/util/heap/monotonepageresource.rs:58]: Ah alright. Sorry I was confused.

**Comment by qinsoon** [comment]
openjdk-microbm
OPENJDK_BINDING_BRANCH_REF=work-packets

File: src/plan/mutator_context.rs
Comment (wenyuzhao) [reply src/plan/mutator_context.rs:25]: What about moving these two functions to `Plan` (as trait methods)? Putting them in mutators is duplication anyway.

Comment (qinsoon) [reply src/plan/mutator_context.rs:25]: Those are preparing/releasing mutators. It feels strange to put them to plan. I think we can just leave them for now.

File: src/plan/gencopy/mutator.rs
Comment (wenyuzhao) [reply src/plan/gencopy/mutator.rs:42]: Done

File: src/scheduler/gc_works.rs
Comment (wenyuzhao) [reply src/scheduler/gc_works.rs]: Fixed. This lock is in BasePlan now.

File: src/scheduler/gc_works.rs
Comment (wenyuzhao) [reply src/scheduler/gc_works.rs:278]: Done.

File: src/scheduler/scheduler.rs
Comment (wenyuzhao) [reply src/scheduler/scheduler.rs:74]: post-alloc initialization is required for now because when MMTK<VM>/Scheduler instance is created, it may not get enough information it required to complete initialization. e.g. the number of worker threads. Not sure how can we fix this...

**Review by qinsoon — APPROVED** [review]
It looks pretty good now. If you would like other members to review this PR, probably try send PMs to let them know that the PR is waiting for their reviews. Otherwise, we should feel confident to merge this PR once the tests pass. For comments that are still unresolved, they mostly are some TODOs - I will add them to an issue when we merge the PR.

File: src/scheduler/scheduler.rs
Comment (wenyuzhao) [reply src/scheduler/scheduler.rs:112]: Fixed. I wrap the coordinator-worker with a mutex and removed all the unsafe code. It'll unlikely hit the performance because the coordinator-worker is rarely used (3-4 times per GC).

File: src/scheduler/scheduler.rs
Comment (wenyuzhao) [reply src/scheduler/scheduler.rs:152]: Fixed

File: src/scheduler/worker.rs
Comment (wenyuzhao) [reply src/scheduler/worker.rs]: Fixed. We don't really need to access the worker-group from a worker. So simply removed this cyclic reference.

File: src/scheduler/work.rs
Comment (wenyuzhao) [reply src/scheduler/work.rs]: Fixed. Now each work is assigned with a unique ID when it is added to a bucket. And the priority-queue in buckets will use this ID for fast work-packet comparison.

File: src/scheduler/work_bucket.rs
Comment (wenyuzhao) [reply src/scheduler/work_bucket.rs]: Fixed.

**Pull request merged**
